# Environment Setup

In [1]:
!nvidia-smi

Sat Sep 19 06:55:05 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   75C    P8             17W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [2]:
!pip install unsloth bitsandbytes peft transformers datasets accelerate trl

In [3]:
import torch, transformers, peft, bitsandbytes, unsloth
print(torch.__version__, transformers.__version__)

/usr/local/lib/python3.13/dist-packages/unsloth/__init__.py:1568: UserWarning: WARNING: Unsloth should be imported before [transformers, peft] to ensure all optimizations are applied. Your code may run slower or encounter memory issues without these optimizations.

Please restructure your imports with 'import unsloth' at the top of your file.
  from ._gpu_init import *


🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
2.11.0+cu128 5.5.0


In [4]:
print(torch.cuda.is_available())

True


# Model Selection

## unsloth/Qwen2.5-Coder-1.5B-Instruct-bnb-4bit

In [ ]:
from unsloth import FastLanguageModel
import torch

In [7]:
model_name = "unsloth/Qwen2.5-Coder-1.5B-Instruct-bnb-4bit"

torch.cuda.reset_peak_memory_stats()
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name=model_name,
    max_seq_length=1024,
    load_in_4bit=True,
)
mem_load_current = torch.cuda.memory_allocated() / 1e9
mem_load_peak = torch.cuda.max_memory_allocated() / 1e9
print(f"{model_name} — after load: current={mem_load_current:.2f} GB, peak={mem_load_peak:.2f} GB")

torch.cuda.reset_peak_memory_stats()
model = FastLanguageModel.get_peft_model(
    model,
    r=16,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"],
    lora_alpha=16,
    lora_dropout=0,
    bias="none",
    use_gradient_checkpointing="unsloth",
    random_state=3407,
)
mem_lora_current = torch.cuda.memory_allocated() / 1e9
mem_lora_peak = torch.cuda.max_memory_allocated() / 1e9
print(f"{model_name} — after LoRA attach: current={mem_lora_current:.2f} GB, peak={mem_lora_peak:.2f} GB")

==((====))==  Unsloth 2026.9.7: Fast Qwen2 patching. Transformers: 5.5.0.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.563 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.11.0+cu128. CUDA: 7.5. CUDA Toolkit: 12.8. Triton: 3.6.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.35. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Loading weights:   0%|          | 0/338 [00:00<?, ?it/s]

unsloth/Qwen2.5-Coder-1.5B-Instruct-bnb-4bit — after load: current=1.19 GB, peak=1.21 GB


Unsloth 2026.9.7 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


unsloth/Qwen2.5-Coder-1.5B-Instruct-bnb-4bit — after LoRA attach: current=1.27 GB, peak=1.27 GB


## unsloth/Qwen2.5-Coder-3B-Instruct-bnb-4bit

In [5]:
from unsloth import FastLanguageModel
import torch

In [6]:
model_name = "unsloth/Qwen2.5-Coder-3B-Instruct-bnb-4bit"

torch.cuda.reset_peak_memory_stats()
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name=model_name,
    max_seq_length=1024,
    load_in_4bit=True,
)
mem_load_current = torch.cuda.memory_allocated() / 1e9
mem_load_peak = torch.cuda.max_memory_allocated() / 1e9
print(f"{model_name} — after load: current={mem_load_current:.2f} GB, peak={mem_load_peak:.2f} GB")

torch.cuda.reset_peak_memory_stats()
model = FastLanguageModel.get_peft_model(
    model,
    r=16,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"],
    lora_alpha=16,
    lora_dropout=0,
    bias="none",
    use_gradient_checkpointing="unsloth",
    random_state=3407,
)
mem_lora_current = torch.cuda.memory_allocated() / 1e9
mem_lora_peak = torch.cuda.max_memory_allocated() / 1e9
print(f"{model_name} — after LoRA attach: current={mem_lora_current:.2f} GB, peak={mem_lora_peak:.2f} GB")

==((====))==  Unsloth 2026.9.7: Fast Qwen2 patching. Transformers: 5.5.0.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.563 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.11.0+cu128. CUDA: 7.5. CUDA Toolkit: 12.8. Triton: 3.6.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.35. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Loading weights:   0%|          | 0/434 [00:00<?, ?it/s]

unsloth/Qwen2.5-Coder-3B-Instruct-bnb-4bit — after load: current=2.10 GB, peak=2.12 GB


Unsloth 2026.9.7 patched 36 layers with 36 QKV layers, 36 O layers and 36 MLP layers.


unsloth/Qwen2.5-Coder-3B-Instruct-bnb-4bit — after LoRA attach: current=2.22 GB, peak=2.22 GB


## unsloth/Llama-3.2-3B-Instruct-bnb-4bit

In [ ]:
from unsloth import FastLanguageModel
import torch

In [6]:
model_name = "unsloth/Llama-3.2-3B-Instruct-bnb-4bit"

torch.cuda.reset_peak_memory_stats()
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name=model_name,
    max_seq_length=1024,
    load_in_4bit=True,
)
mem_load_current = torch.cuda.memory_allocated() / 1e9
mem_load_peak = torch.cuda.max_memory_allocated() / 1e9
print(f"{model_name} — after load: current={mem_load_current:.2f} GB, peak={mem_load_peak:.2f} GB")

torch.cuda.reset_peak_memory_stats()
model = FastLanguageModel.get_peft_model(
    model,
    r=16,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"],
    lora_alpha=16,
    lora_dropout=0,
    bias="none",
    use_gradient_checkpointing="unsloth",
    random_state=3407,
)
mem_lora_current = torch.cuda.memory_allocated() / 1e9
mem_lora_peak = torch.cuda.max_memory_allocated() / 1e9
print(f"{model_name} — after LoRA attach: current={mem_lora_current:.2f} GB, peak={mem_lora_peak:.2f} GB")

==((====))==  Unsloth 2026.9.7: Fast Llama patching. Transformers: 5.5.0.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.563 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.11.0+cu128. CUDA: 7.5. CUDA Toolkit: 12.8. Triton: 3.6.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.35. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Loading weights:   0%|          | 0/254 [00:00<?, ?it/s]

Unsloth: Will load unsloth/Llama-3.2-3B-Instruct-bnb-4bit as a legacy tokenizer.


unsloth/Llama-3.2-3B-Instruct-bnb-4bit — after load: current=2.29 GB, peak=2.31 GB


Unsloth 2026.9.7 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


unsloth/Llama-3.2-3B-Instruct-bnb-4bit — after LoRA attach: current=2.39 GB, peak=2.39 GB


## Comparison

In [7]:
import pandas as pd

results = [
    {"model": "Qwen2.5-Coder-1.5B", "mem_load_gb": 1.21, "mem_lora_gb": 1.27},
    {"model": "Qwen2.5-Coder-3B",   "mem_load_gb": 2.12, "mem_lora_gb": 2.22},
    {"model": "Llama-3.2-3B",       "mem_load_gb": 2.31, "mem_lora_gb": 2.39},
]

df = pd.DataFrame(results)
df["headroom_gb"] = 15 - df["mem_lora_gb"]
df

,model,mem_load_gb,mem_lora_gb,headroom_gb
0,Qwen2.5-Coder-1.5B,1.21,1.27,13.73
1,Qwen2.5-Coder-3B,2.12,2.22,12.78
2,Llama-3.2-3B,2.31,2.39,12.61


Selecting unsloth/Qwen2.5-Coder-3B-Instruct-bnb-4bit

# Dataset Exploration

In [7]:
!apt-get install -y git-lfs
!git lfs install

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
git-lfs is already the newest version (3.4.1-1ubuntu0.4).
0 upgraded, 0 newly installed, 0 to remove and 52 not upgraded.
Git LFS initialized.


In [8]:
%cd spider-data
!git lfs pull
%cd ..

/content/spider-data
/content


In [9]:
!git lfs install
!git clone https://huggingface.co/datasets/minktn/spider-data
!ls spider-data

Git LFS initialized.
fatal: destination path 'spider-data' already exists and is not an empty directory.
README.md  spider_data


In [10]:
from datasets import load_dataset

spider = load_dataset("xlangai/spider")
print(spider)

DatasetDict({
    train: Dataset({
        features: ['db_id', 'query', 'question', 'query_toks', 'query_toks_no_value', 'question_toks'],
        num_rows: 7000
    })
    validation: Dataset({
        features: ['db_id', 'query', 'question', 'query_toks', 'query_toks_no_value', 'question_toks'],
        num_rows: 1034
    })
})


In [11]:
example = spider["train"][0]
print(example)

{'db_id': 'department_management', 'query': 'SELECT count(*) FROM head WHERE age  >  56', 'question': 'How many heads of the departments are older than 56 ?', 'query_toks': ['SELECT', 'count', '(', '*', ')', 'FROM', 'head', 'WHERE', 'age', '>', '56'], 'query_toks_no_value': ['select', 'count', '(', '*', ')', 'from', 'head', 'where', 'age', '>', 'value'], 'question_toks': ['How', 'many', 'heads', 'of', 'the', 'departments', 'are', 'older', 'than', '56', '?']}


In [12]:
import json

with open("spider-data/spider_data/tables.json") as f:
    tables = json.load(f)

schema_lookup = {t["db_id"]: t for t in tables}

db_id = example["db_id"]
schema = schema_lookup[db_id]

print("db_id:", db_id)
print("table_names:", schema["table_names"])
print("column_names:", schema["column_names"][:10])  # first 10 for brevity
print("column_types:", schema["column_types"][:10])

db_id: department_management
table_names: ['department', 'head', 'management']
column_names: [[-1, '*'], [0, 'department id'], [0, 'name'], [0, 'creation'], [0, 'ranking'], [0, 'budget in billions'], [0, 'num employees'], [1, 'head id'], [1, 'name'], [1, 'born state']]
column_types: ['text', 'number', 'text', 'text', 'number', 'number', 'number', 'number', 'text', 'text']


In [13]:
import os

sqlite_path = f"spider-data/spider_data/database/{db_id}/{db_id}.sqlite"
print(os.path.exists(sqlite_path), sqlite_path)

True spider-data/spider_data/database/department_management/department_management.sqlite


In [14]:
print("Question:", example["question"])
print("Gold SQL:", example["query"])
print("DB:", db_id)
print("Tables:", schema["table_names"])

Question: How many heads of the departments are older than 56 ?
Gold SQL: SELECT count(*) FROM head WHERE age  >  56
DB: department_management
Tables: ['department', 'head', 'management']


Each Spider example has a db_id linking it to one specific database.
tables.json holds the schema (tables, columns, types, foreign keys) for every db_id.
The actual data lives in a separate .sqlite file per db_id, used later for execution accuracy.
So one example = (question, gold SQL, db_id) + schema lookup + .sqlite file, all joined on db_id.

# Preprocessing

In [15]:
def schema_to_text(db_id, schema_lookup):
    schema = schema_lookup[db_id]
    table_names = schema["table_names_original"]
    column_names = schema["column_names_original"]  # [(table_idx, col_name), ...]
    column_types = schema["column_types"]
    pk_cols = set(schema.get("primary_keys", []))
    fk_pairs = schema.get("foreign_keys", [])

    tables = {i: [] for i in range(len(table_names))}
    for col_idx, (tbl_idx, col_name) in enumerate(column_names):
        if tbl_idx == -1:
            continue
        col_type = column_types[col_idx]
        pk_marker = " PRIMARY KEY" if col_idx in pk_cols else ""
        tables[tbl_idx].append(f"{col_name} {col_type}{pk_marker}")

    lines = []
    for tbl_idx, tbl_name in enumerate(table_names):
        cols = ", ".join(tables[tbl_idx])
        lines.append(f"CREATE TABLE {tbl_name} ({cols});")

    if fk_pairs:
        fk_lines = []
        for from_col_idx, to_col_idx in fk_pairs:
            from_tbl, from_col = column_names[from_col_idx]
            to_tbl, to_col = column_names[to_col_idx]
            fk_lines.append(
                f"-- FK: {table_names[from_tbl]}.{from_col} -> {table_names[to_tbl]}.{to_col}"
            )
        lines.extend(fk_lines)

    return "\n".join(lines)

In [16]:
def build_prompt(schema_text, question, gold_sql=None):
    system_msg = (
        "You are a text-to-SQL model. Given a database schema and a question, "
        "output only the SQL query. Do not include explanations or markdown formatting."
    )
    user_msg = f"Schema:\n{schema_text}\n\nQuestion: {question}"

    messages = [
        {"role": "system", "content": system_msg},
        {"role": "user", "content": user_msg},
    ]

    if gold_sql is not None:
        # training version: include the assistant turn with gold SQL
        messages.append({"role": "assistant", "content": gold_sql})

    return messages

In [17]:
from transformers import AutoTokenizer
# Inference version: no gold SQL, add_generation_prompt=True so the model knows to respond next
inference_messages = build_prompt(schema_to_text(db_id, schema_lookup), example["question"])
inference_prompt = tokenizer.apply_chat_template(
    inference_messages, tokenize=False, add_generation_prompt=True
)
print("=== INFERENCE PROMPT ===")
print(inference_prompt)

# Training version: gold SQL included as the assistant turn, no generation prompt needed
training_messages = build_prompt(
    schema_to_text(db_id, schema_lookup), example["question"], gold_sql=example["query"]
)
training_prompt = tokenizer.apply_chat_template(
    training_messages, tokenize=False, add_generation_prompt=False
)
print("=== TRAINING PROMPT ===")
print(training_prompt)

=== INFERENCE PROMPT ===
<|im_start|>system
You are a text-to-SQL model. Given a database schema and a question, output only the SQL query. Do not include explanations or markdown formatting.<|im_end|>
<|im_start|>user
Schema:
CREATE TABLE department (Department_ID number PRIMARY KEY, Name text, Creation text, Ranking number, Budget_in_Billions number, Num_Employees number);
CREATE TABLE head (head_ID number PRIMARY KEY, name text, born_state text, age number);
CREATE TABLE management (department_ID number PRIMARY KEY, head_ID number, temporary_acting text);
-- FK: management.head_ID -> head.head_ID
-- FK: management.department_ID -> department.Department_ID

Question: How many heads of the departments are older than 56 ?<|im_end|>
<|im_start|>assistant

=== TRAINING PROMPT ===
<|im_start|>system
You are a text-to-SQL model. Given a database schema and a question, output only the SQL query. Do not include explanations or markdown formatting.<|im_end|>
<|im_start|>user
Schema:
CREATE TA

In [18]:
for i in range(5):
    ex = spider["train"][i]
    db_id_i = ex["db_id"]
    schema_text = schema_to_text(db_id_i, schema_lookup)
    msgs = build_prompt(schema_text, ex["question"])
    rendered = tokenizer.apply_chat_template(msgs, tokenize=False, add_generation_prompt=True)
    print(f"--- Example {i} (db_id={db_id_i}) ---")
    print(rendered[:800])  # truncate for readability while scanning
    print("=======================================================")

--- Example 0 (db_id=department_management) ---
<|im_start|>system
You are a text-to-SQL model. Given a database schema and a question, output only the SQL query. Do not include explanations or markdown formatting.<|im_end|>
<|im_start|>user
Schema:
CREATE TABLE department (Department_ID number PRIMARY KEY, Name text, Creation text, Ranking number, Budget_in_Billions number, Num_Employees number);
CREATE TABLE head (head_ID number PRIMARY KEY, name text, born_state text, age number);
CREATE TABLE management (department_ID number PRIMARY KEY, head_ID number, temporary_acting text);
-- FK: management.head_ID -> head.head_ID
-- FK: management.department_ID -> department.Department_ID

Question: How many heads of the departments are older than 56 ?<|im_end|>
<|im_start|>assistant

--- Example 1 (db_id=department_management) ---
<|im_start|>system
You are a text-to-SQL model. Given a database schema and a question, output only the SQL query. Do not include explanations or markdown formattin

## Train/Tests Split

In [19]:
import random

train_pool = spider["train"]
test_set = spider["validation"]

print("Train pool size:", len(train_pool))
print("Test set size:", len(test_set))

Train pool size: 7000
Test set size: 1034


In [20]:
SEED = 42
random.seed(SEED)

n = len(train_pool)
indices = list(range(n))
random.shuffle(indices)

val_fraction = 0.08  # ~8%, within the PRD's 5-10% guidance
val_size = int(n * val_fraction)

val_indices = set(indices[:val_size])
train_indices = [i for i in indices if i not in val_indices]

train_split_raw = train_pool.select(train_indices)
val_split_raw = train_pool.select(sorted(val_indices))

print("Train:", len(train_split_raw))
print("Val:", len(val_split_raw))
print("Test:", len(test_set))

Train: 6440
Val: 560
Test: 1034


In [21]:
def format_example(ex, is_training):
    db_id_i = ex["db_id"]
    schema_text = schema_to_text(db_id_i, schema_lookup)
    gold_sql = ex["query"] if is_training else None
    messages = build_prompt(schema_text, ex["question"], gold_sql=gold_sql)
    add_gen_prompt = not is_training

    rendered = tokenizer.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=add_gen_prompt
    )

    return {
        "db_id": db_id_i,
        "question": ex["question"],
        "gold_sql": ex["query"],
        "schema_text": schema_text,
        "formatted_prompt": rendered,
    }

def format_split(split, is_training):
    return [format_example(ex, is_training) for ex in split]

train_formatted = format_split(train_split_raw, is_training=True)
val_formatted = format_split(val_split_raw, is_training=True)
test_formatted = format_split(test_set, is_training=False)  # inference-mode: no gold SQL baked into prompt

print("Formatted train:", len(train_formatted))
print("Formatted val:", len(val_formatted))
print("Formatted test:", len(test_formatted))

Formatted train: 6440
Formatted val: 560
Formatted test: 1034


In [22]:
from google.colab import drive
drive.mount('/content/drive')

import json, os

save_dir = "/content/drive/MyDrive/text-to-sql-finetuning/data"
os.makedirs(save_dir, exist_ok=True)

def save_jsonl(data, path):
    with open(path, "w") as f:
        for row in data:
            f.write(json.dumps(row) + "\n")

save_jsonl(train_formatted, f"{save_dir}/train.jsonl")
save_jsonl(val_formatted, f"{save_dir}/val.jsonl")
save_jsonl(test_formatted, f"{save_dir}/test.jsonl")

print("Saved to", save_dir)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Saved to /content/drive/MyDrive/text-to-sql-finetuning/data


In [23]:
print("Train rows:", len(train_formatted))
print("Val rows:", len(val_formatted))
print("Test rows:", len(test_formatted))

for i in [0, 1]:
    ex = test_formatted[i]
    print(f"\n--- Test example {i} ---")
    print("db_id:", ex["db_id"])
    print("question:", ex["question"])
    print("gold_sql:", ex["gold_sql"])
    print("schema_text present:", bool(ex["schema_text"]))

Train rows: 6440
Val rows: 560
Test rows: 1034

--- Test example 0 ---
db_id: concert_singer
question: How many singers do we have?
gold_sql: SELECT count(*) FROM singer
schema_text present: True

--- Test example 1 ---
db_id: concert_singer
question: What is the total number of singers?
gold_sql: SELECT count(*) FROM singer
schema_text present: True


In [24]:
len(test_formatted) == len(spider["validation"])

True

In [25]:
assert len(set(train_indices) & val_indices) == 0
print("No overlap between train and val — confirmed")

No overlap between train and val — confirmed


# SQL Lite DB

In [26]:
import sqlite3
import os

SPIDER_DB_ROOT = "spider-data/spider_data/database"  # matches the nested path you found earlier

def get_db_connection(db_id):
    db_path = os.path.join(SPIDER_DB_ROOT, db_id, f"{db_id}.sqlite")
    if not os.path.exists(db_path):
        raise FileNotFoundError(f"No sqlite file found for db_id={db_id} at {db_path}")

    # open read-only so eval runs can't accidentally mutate the shared file
    uri = f"file:{db_path}?mode=ro"
    conn = sqlite3.connect(uri, uri=True)
    return conn

In [27]:
def run_sql(conn, sql_string):
    try:
        cursor = conn.cursor()
        cursor.execute(sql_string)
        rows = cursor.fetchall()
        return True, rows, None
    except Exception as e:
        return False, [], str(e)

In [28]:
example = spider["train"][0]
db_id = example["db_id"]
gold_sql = example["query"]

conn = get_db_connection(db_id)
success, rows, error = run_sql(conn, gold_sql)

print("Success:", success)
print(db_id)
print("Num rows:", len(rows))
print("First few rows:", rows[:5])
print("Error:", error)

Success: True
department_management
Num rows: 1
First few rows: [(5,)]
Error: None


In [29]:
broken_sql = "SELECT * FROM this_table_does_not_exist;"
success, rows, error = run_sql(conn, broken_sql)

print("Success:", success)   # should be False
print("Rows:", rows)          # should be []
print("Error:", error)        # should have a real sqlite error message

Success: False
Rows: []
Error: no such table: this_table_does_not_exist


In [30]:
!find spider_data_clone -iname "*.sqlite" | head
!find spider_data_clone -iname "*.db" | head

In [31]:
import os

# adjust this path once Cell 1 shows you the actual structure
candidate_path = "spider-data/spider_data/database/department_management/department_management.sqlite"

print("Exists:", os.path.exists(candidate_path))
if os.path.exists(candidate_path):
    with open(candidate_path, "rb") as f:
        header = f.read(16)
    print("Header:", header)  # should be b'SQLite format 3\x00'

Exists: True
Header: b'SQLite format 3\x00'


In [32]:
mutate_sql = f"DELETE FROM {schema_lookup[db_id]['table_names_original'][0]};"
success, rows, error = run_sql(conn, mutate_sql)

print("Success:", success)   # should be False
print("Error:", error)        # should mention read-only

Success: False
Error: attempt to write a readonly database


# Accuracy and Exacat match Comparison func

In [33]:
def execution_match(conn, generated_sql, gold_sql):
    gen_success, gen_rows, gen_error = run_sql(conn, generated_sql)
    gold_success, gold_rows, gold_error = run_sql(conn, gold_sql)

    if not gen_success or not gold_success:
        return False

    # Policy: rows compared as sets of tuples (order-insensitive),
    # columns compared positionally within each row (not sorted) —
    # i.e. two rows match only if their values line up in the same column order,
    # but the overall set of rows can appear in any order.
    gen_set = set(tuple(row) for row in gen_rows)
    gold_set = set(tuple(row) for row in gold_rows)

    return gen_set == gold_set

In [34]:
def exact_match(generated_sql, gold_sql):
    def normalize(sql):
        return sql.strip().rstrip(";").strip().lower()

    # NOTE: this is a weak secondary metric — it penalizes any valid
    # rephrasing (different column order, different but equivalent JOIN syntax,
    # extra whitespace inside the query, alias naming differences, etc.)
    # even when execution_match would correctly call it correct.
    return normalize(generated_sql) == normalize(gold_sql)

In [35]:
test_sql_a = "SELECT name FROM department;"
print("execution_match:", execution_match(conn, test_sql_a, test_sql_a))  # expect True
print("exact_match:", exact_match(test_sql_a, test_sql_a))                # expect True

execution_match: True
exact_match: True


In [36]:
test_sql_b1 = "SELECT name, budget FROM department;"
test_sql_b2 = "SELECT budget, name FROM department;"

print("execution_match:", execution_match(conn, test_sql_b1, test_sql_b2))
# ^ expect False under our "positional column comparison" policy —
#   the columns are swapped, so tuples won't match position-for-position
#   even though a human would call these "the same information differently ordered"
print("exact_match:", exact_match(test_sql_b1, test_sql_b2))  # expect False

execution_match: False
exact_match: False


In [37]:
test_sql_c = "SELECT COUNT(*) FROM department;"

print("execution_match:", execution_match(conn, test_sql_b1, test_sql_c))  # expect False
print("exact_match:", exact_match(test_sql_b1, test_sql_c))                # expect False

execution_match: False
exact_match: False


Comparison policy: rows compared as sets of tuples (order-insensitive across rows),
columns compared positionally (not sorted within a row) — meaning a query that returns
the same values in a different column order will NOT count as a match. This is a
deliberate tradeoff: positional comparison avoids false-positive matches between
queries that coincidentally share values across different columns.

# Eval Harness

In [52]:
from tqdm.notebook import tqdm
import json

def run_eval_harness(test_examples, generate_fn, results_path="eval_results.json"):
    per_example_results = []
    n_exec_match = 0
    n_exact_match = 0

    for i, ex in enumerate(tqdm(test_examples, desc="Evaluating")):
        db_id = ex["db_id"]
        question = ex["question"]
        gold_sql = ex["gold_sql"]
        schema_text = ex["schema_text"]

        try:
            generated_sql = generate_fn(question, schema_text)
        except Exception as e:
            generated_sql = ""
            print(f"[{i}] generate_fn raised an exception: {e}")

        conn = get_db_connection(db_id)

        exec_ok = execution_match(conn, generated_sql, gold_sql)
        exact_ok = exact_match(generated_sql, gold_sql)

        if exec_ok:
            n_exec_match += 1
        if exact_ok:
            n_exact_match += 1

        per_example_results.append({
            "index": i,
            "db_id": db_id,
            "question": question,
            "generated_sql": generated_sql,
            "gold_sql": gold_sql,
            "execution_match": exec_ok,
            "exact_match": exact_ok,
        })

    n = len(test_examples)
    summary = {
        "execution_accuracy": n_exec_match / n if n > 0 else 0.0,
        "exact_match": n_exact_match / n if n > 0 else 0.0,
        "n": n,
    }

    with open(results_path, "w") as f:
        json.dump({"summary": summary, "per_example": per_example_results}, f, indent=2)

    print("Summary:", summary)
    print(f"Per-example results saved to {results_path}")

    return {"summary": summary, "per_example": per_example_results}

In [53]:
# Use a small subset of test_formatted for a fast sanity check
sanity_subset = test_formatted[:20]

def perfect_generate_fn(question, schema_text):
    # cheat: find the matching gold SQL from the subset by question text
    for ex in sanity_subset:
        if ex["question"] == question:
            return ex["gold_sql"]
    return ""

results_perfect = run_eval_harness(sanity_subset, perfect_generate_fn, results_path="eval_perfect_sanity.json")

Evaluating:   0%|          | 0/20 [00:00<?, ?it/s]

Summary: {'execution_accuracy': 1.0, 'exact_match': 1.0, 'n': 20}
Per-example results saved to eval_perfect_sanity.json


In [54]:
with open("eval_perfect_sanity.json") as f:
    loaded = json.load(f)

print(loaded["summary"])
print(loaded["per_example"][0])

{'execution_accuracy': 1.0, 'exact_match': 1.0, 'n': 20}
{'index': 0, 'db_id': 'concert_singer', 'question': 'How many singers do we have?', 'generated_sql': 'SELECT count(*) FROM singer', 'gold_sql': 'SELECT count(*) FROM singer', 'execution_match': True, 'exact_match': True}


## Testing Eval Harness

In [55]:
def broken_generate_fn(question, schema_text):
    return "SELECT 1;"

results_broken = run_eval_harness(sanity_subset, broken_generate_fn, results_path="eval_broken_sanity.json")

Evaluating:   0%|          | 0/20 [00:00<?, ?it/s]

Summary: {'execution_accuracy': 0.0, 'exact_match': 0.0, 'n': 20}
Per-example results saved to eval_broken_sanity.json


In [56]:
print("Perfect generator:", results_perfect["summary"])
print("Broken generator:", results_broken["summary"])

Perfect generator: {'execution_accuracy': 1.0, 'exact_match': 1.0, 'n': 20}
Broken generator: {'execution_accuracy': 0.0, 'exact_match': 0.0, 'n': 20}


In [57]:
# Search the perfect-generator results for a case where gold and generated differ in
# formatting/order but still execute to the same result — here we construct one manually
# since the perfect generator returns gold_sql verbatim (so exact_match is trivially True there).
# Instead, build a hand-picked example: take one gold query and rewrite it in a logically
# equivalent but textually different form.

example_gold = test_formatted[0]["gold_sql"]
db_id_0 = test_formatted[0]["db_id"]
print("db_id:", db_id_0)
print("Original gold SQL:", example_gold)

# Manually construct a reformatted-but-equivalent version, e.g. reordering a WHERE clause,
# adding explicit semicolon/whitespace differences, or aliasing — pick based on what
# example_gold actually looks like. Example pattern if it's a simple SELECT:
# reformatted = example_gold.replace("SELECT", "select").strip() + "   "

# Once you have a real reformatted version, test both:
reformatted_sql = "select   COUNT(*)   from singer;"

db_id_0 = test_formatted[0]["db_id"]
conn0 = get_db_connection(db_id_0)
print("Reformatted SQL:", reformatted_sql)
print("execution_match:", execution_match(conn0, reformatted_sql, example_gold))
print("exact_match:", exact_match(reformatted_sql, example_gold))

db_id: concert_singer
Original gold SQL: SELECT count(*) FROM singer
Reformatted SQL: select   COUNT(*)   from singer;
execution_match: True
exact_match: False


Sanity check 1: "perfect" generator (returns gold SQL unchanged) — execution accuracy ≈ 100%
Sanity check 2: "broken" generator (always returns "SELECT 1;") — execution accuracy ≈ 0%
These two extremes confirm the harness correctly distinguishes good SQL from bad SQL,
not just that it runs without crashing.

Exact-match vs execution-match example:
  Gold:        SELECT count(*) FROM singer
  Reformatted: select   COUNT(*)   from singer;
  execution_match: True
  exact_match: False
This demonstrates that exact_match penalizes valid textual variation, while
execution_match correctly recognizes logically equivalent queries.

Design decision made deliberately: columns are compared positionally (not sorted)
within each matched row — see Task 7 notes for the tradeoff reasoning.

## Zero-shot Testing


In [58]:
import re

def extract_sql(raw_output):
    """Strip chat template artifacts, markdown fences, and explanation text
    down to just the SQL query."""
    text = raw_output.strip()

    # strip markdown code fences if present
    fence_match = re.search(r"```(?:sql)?\s*(.*?)```", text, re.DOTALL)
    if fence_match:
        text = fence_match.group(1).strip()

    # if the model added explanation before/after, try to isolate the first
    # SQL-looking statement (starts with SELECT/INSERT/UPDATE/DELETE/WITH)
    sql_start_match = re.search(
        r"(SELECT|WITH|INSERT|UPDATE|DELETE)\b", text, re.IGNORECASE
    )
    if sql_start_match:
        text = text[sql_start_match.start():]

    # cut off at the first semicolon if present, else take the whole remainder
    if ";" in text:
        text = text.split(";")[0] + ";"

    return text.strip()


def generate_sql(question, schema_text, model, tokenizer, max_new_tokens=256):
    messages = build_prompt(schema_text, question)  # inference mode: no gold_sql
    prompt = tokenizer.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=True
    )

    inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

    outputs = model.generate(
        **inputs,
        max_new_tokens=max_new_tokens,
        do_sample=False,   # greedy decoding for reproducibility
        pad_token_id=tokenizer.eos_token_id,
    )

    generated_ids = outputs[0][inputs["input_ids"].shape[1]:]  # only new tokens
    raw_text = tokenizer.decode(generated_ids, skip_special_tokens=True)

    return extract_sql(raw_text)

In [59]:
def baseline_generate_fn(question, schema_text):
    return generate_sql(question, schema_text, model, tokenizer)

In [68]:
import random

random.seed(42)
SUBSET_SIZE = 200

if len(test_formatted) > SUBSET_SIZE:
    test_subset = random.sample(test_formatted, SUBSET_SIZE)
else:
    test_subset = test_formatted

print(f"Running baseline on {len(test_subset)} examples (full test set: {len(test_formatted)})")

Running baseline on 200 examples (full test set: 1034)


In [69]:
import warnings
warnings.filterwarnings("ignore", message=".*max_new_tokens.*")

In [70]:
import time

start = time.time()
baseline_results = run_eval_harness(
    test_subset,
    baseline_generate_fn,
    results_path="baseline_zero_shot_results.json",
)
elapsed = time.time() - start

print(f"Completed {len(test_subset)} examples in {elapsed:.1f}s "
      f"({elapsed / len(test_subset):.2f}s/example)")

Evaluating:   0%|          | 0/200 [00:00<?, ?it/s]

Both `max_new_tokens` (=256) and `max_length`(=32768) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=256) and `max_length`(=32768) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=256) and `max_length`(=32768) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=256) and `max_length`(=32768) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_gene

Summary: {'execution_accuracy': 0.61, 'exact_match': 0.105, 'n': 200}
Per-example results saved to baseline_zero_shot_results.json
Completed 200 examples in 636.8s (3.18s/example)


In [71]:
failures = [r for r in baseline_results["per_example"] if not r["execution_match"]]
print(f"Total failures: {len(failures)} / {len(test_subset)}")

for f in failures[:10]:
    print("=" * 60)
    print("Question:", f["question"])
    print("Generated:", f["generated_sql"])
    print("Gold:", f["gold_sql"])

Total failures: 78 / 200
Question: Give the code of the airport with the least flights.
Generated: SELECT AirportCode FROM airports ORDER BY COUNT(*) ASC LIMIT 1
Gold: SELECT T1.AirportCode FROM AIRPORTS AS T1 JOIN FLIGHTS AS T2 ON T1.AirportCode  =  T2.DestAirport OR T1.AirportCode  =  T2.SourceAirport GROUP BY T1.AirportCode ORDER BY count(*) LIMIT 1
Question: What are the distinct battle names which are between bulgarian commander 'Kaloyan' and latin commander 'Baldwin I'?
Generated: SELECT DISTINCT T1.name FROM battle AS T1 INNER JOIN ship AS T2 ON T1.id = T2.lost_in_battle WHERE T1.bulgarian_commander = 'Kaloyan' AND T1.latin_commander = 'Baldwin I'
Gold: SELECT name FROM battle WHERE bulgarian_commander  =  'Kaloyan' AND latin_commander  =  'Baldwin I'
Question: What are the ids and names of all countries that either have more than 3 car makers or produce fiat model ?
Generated: SELECT T2.CountryId, T1.CountryName FROM countries AS T1 INNER JOIN car_makers AS T2 ON T1.CountryId =

In [72]:
print("=== Zero-shot baseline (Qwen2.5-Coder-3B) ===")
print(f"Test subset size: {len(test_subset)}")
print(f"Execution accuracy: {baseline_results['summary']['execution_accuracy']:.2%}")
print(f"Exact match: {baseline_results['summary']['exact_match']:.2%}")

=== Zero-shot baseline (Qwen2.5-Coder-3B) ===
Test subset size: 200
Execution accuracy: 61.00%
Exact match: 10.50%


In [73]:
with open("baseline_zero_shot_results.json") as f:
    loaded = json.load(f)
print(loaded["summary"])

{'execution_accuracy': 0.61, 'exact_match': 0.105, 'n': 200}


## Findings — Text-to-SQL Baseline

**Base model:** Qwen2.5-Coder-3B-Instruct (4-bit) — chosen over Qwen2.5-Coder-1.5B and Llama-3.2-3B for stronger code/structured-output priors, with ~12.8GB headroom on the T4 (see Task 2 table).

**Dataset:** Spider — train pool split 92/8 into train/val (seed=42), dev set used as test (unmodified, per PRD).

**Test subset:** 200 examples (fixed seed=42 sample of Spider dev), greedy decoding.

**Baseline results:**
- Execution accuracy: **61.0%** (122/200)
- Exact match: *[insert from `loaded["summary"]["exact_match"]`]*

**Recurring failure patterns (from manual review of 10 failures):**
1. **Unnecessary/hallucinated joins** — most common pattern. Model joins to unrelated tables (e.g. `country` → `countrylanguage` for a query needing only `country`'s own columns) instead of checking if needed columns already sit on one table.
2. **Self-join failures** — e.g. "Kyle's friends" needed `Highschooler` joined to itself twice (once per role); model only joined once, returning Kyle instead of his friends. Known hard case for text-to-SQL.
3. **Set operations replaced with JOIN/WHERE logic** — `UNION`/`INTERSECT` queries rewritten as single joins with `OR`, breaking aggregate semantics.
4. **`!=` vs `NOT IN` (subquery) confusion** — filters individual rows instead of excluding entire entities.

**Difficulty breakdown:** *not performed — Spider's difficulty labels weren't used for this pass* (noted per completion criteria rather than skipped silently).

**Takeaway:** 61% execution accuracy leaves clear room for fine-tuning to move the needle, and the join-hallucination pattern is a good target for Phase 2 error-analysis-driven data curation.